Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/TapToPlace.cs

```diff
@@ -0,0 +1,373 @@
+        private LayerMask[] magneticSurfaces = { UnityEngine.Physics.DefaultRaycastLayers };
+
+        /// <summary>
+        /// Array of LayerMask to execute from highest to lowest priority. First layermask to provide a raycast hit will be used by component.
+        /// </summary>
+        public LayerMask[] MagneticSurfaces
+        {
+            get => magneticSurfaces;
+            set => magneticSurfaces = value;
+        }
+
+        [SerializeField]
+        [Tooltip("If true and in the Unity Editor, the normal of the raycast hit will be drawn in yellow.")]
+        private bool debugEnabled = true;
+
+        /// <summary>
+        /// If true and in the Unity Editor, the normal of the raycast hit will be drawn in yellow.
+        /// </summary>
+        public bool DebugEnabled
+        {
+            get => debugEnabled;
+            set => debugEnabled = value;
+        }
+
+        [SerializeField, FormerlySerializedAs("OnPlacingStarted")]
+        [Tooltip("This event is triggered once when the game object to place is selected.")]
+        private UnityEvent onPlacingStarted = new UnityEvent();
+
+        /// <summary>
+        /// This event is triggered once when the game object to place is selected.
+        /// </summary>
+        public UnityEvent OnPlacingStarted
+        {
+            get => onPlacingStarted;
+            set => onPlacingStarted = value;
+        }
+
+        [SerializeField, FormerlySerializedAs("OnPlacingStopped")]
+        [Tooltip("This event is triggered once when the game object to place is unselected, placed.")]
+        private UnityEvent onPlacingStopped = new UnityEvent();
+
+        /// <summary>
+        /// This event is triggered once when the game object to place is unselected, placed.
+        /// </summary>
+        public UnityEvent OnPlacingStopped
+        {
+            get => onPlacingStopped;
+            set => onPlacingStopped = value;
+        }
+
+        protected internal int GameObjectLayer { get; protected set; }
+
+        protected internal bool IsColliderPresent => gameObject != null ? gameObject.GetComponent<Collider>() != null : false;
+
+        private int ignoreRaycastLayer;
+
+        // The current ray is based on the TrackedTargetType (Controller Ray, Head, Hand Joint)
+        protected RayStep CurrentRay;
+
+        protected bool DidHitSurface;
+
+        protected RaycastHit CurrentHit;
+
+        protected float LastTimeClicked = 0;
+
+        protected float doubleClickTimeout = 0.5f;
+
+        #region MonoBehaviour Implementation
+        protected override void Start()
+        {
+            // Solver base class
+            base.Start();
+
+            Debug.Assert(IsColliderPresent, "The game object " + gameObject.name + " does not have a collider attached, please attach a collider to use Tap to Place");
```

Comment:
Can write it like this:

```suggestion
            Debug.Assert(IsColliderPresent, $"The game object {gameObject.name} does not have a collider attached, please attach a collider to use Tap to Place");
```

Reply:
Double check this compiles, I might have made a typo.

Reply:
Just checked and it compiles.

Reply:
FWIW I think that string interpolation like this is generally nicer - makes for less " + ... + " things.